### PR TITLE
make Userwhitelist plugin generic

### PR DESCRIPTION
**Pull request by vladak**
This change allows to choose the field to check against in the `Userwhitelist` plugin.

### CODE REVIEW

File: plugins/src/main/java/opengrok/auth/plugin/UserWhiteListPlugin.java [hunk 1]
```diff
@@ -77,7 +90,13 @@ private boolean checkWhitelist(HttpServletRequest request) {
             return false;
         }
 
-        return whitelist.contains(user.getUsername());
+        if (fieldName.equals(USERNAME_FIELD)) {
+            return user.getUsername() != null && whitelist.contains(user.getUsername());
+        } else if (fieldName.equals(ID_FIELD)) {
+            return user.getId() != null && whitelist.contains(user.getId());
+        }
```
Comment (tulinkry) [inline plugins/src/main/java/opengrok/auth/plugin/UserWhiteListPlugin.java:97]: I'd prefer enum, should be faster, could be arguably more readable

File: plugins/src/main/java/opengrok/auth/plugin/UserWhiteListPlugin.java [hunk 2]
```diff
@@ -44,18 +44,31 @@
...
+    // configuration parameters
     static final String FILE_PARAM = "file";
+    static final String FIELD_PARAM = "fieldName";
+
+    // valid values for the FIELD_NAME parameter
+    static final String USERNAME_FIELD = "username";
+    static final String ID_FIELD = "id";
 
     private final Set<String> whitelist = new TreeSet<>();
+    private String fieldName;
```
Comment (tulinkry) [inline plugins/src/main/java/opengrok/auth/plugin/UserWhiteListPlugin.java:56]: can be final

File: plugins/src/main/java/opengrok/auth/plugin/UserWhiteListPlugin.java [hunk 3]
```diff
@@ -18,7 +18,7 @@
  */
 
 /*
- * Copyright (c) 2019 Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2019-2020 Oracle and/or its affiliates. All rights reserved.
```
Comment (ahornace) [inline plugins/src/main/java/opengrok/auth/plugin/UserWhiteListPlugin.java:21]: Aren't we using `2019, 2020,` format?

Comment (vladak) [reply plugins/src/main/java/opengrok/auth/plugin/UserWhiteListPlugin.java:21]: yes, fixed. There should be a copyright format checker I guess. Filed #3300.

File: plugins/src/main/java/opengrok/auth/plugin/UserWhiteListPlugin.java (hunk 2 above)
Comment (vladak) [reply plugins/src/main/java/opengrok/auth/plugin/UserWhiteListPlugin.java:56]: It cannot. The value is assigned in `load()`.

File: plugins/src/main/java/opengrok/auth/plugin/UserWhiteListPlugin.java (hunk 1 above)
Comment (vladak) [reply plugins/src/main/java/opengrok/auth/plugin/UserWhiteListPlugin.java:97]: I'd rather keep the configuration parameters of plugins restricted to strings. Or do you mean introduce the enum alongside the configuration only for internal handling in the plugin ?

**Comment by vladak** [comment]
I have to say it was definitely pleasant to have the test already in place, @idodeclare. Increasing test coverage definitely makes sense.

**Comment by vladak** [comment]
will update https://github.com/oracle/opengrok/wiki/Authorization-plugins#userwhitelistplugin

**Pull request merged**